### ISSUE TITLE
How to run this?

### ISSUE THREAD

**Issue by DanielFabian** [Trundle/NixOS-WSL#1]
I've created the rootfs using your nix expression. But I don't really know how to launch it next?

I didn't go through the full tutorial about how to package a distro, but just used WSL2's `import` command. After that, I can launch the distro, It only shows `-sh-4.4#`, but when I access some of the commands at least **something** happens. But I don't really know how to get it to the point where I can `nixos-install` or `nixos-rebuild`. A bit of help would be very much appreciated.

**Comment by Trundle** [comment]
The system profile needs to be activated first (only once): `/nix/var/nix/profiles/system/activate`

I installed the root FS using a fork of [WSL DistroLauncher](https://github.com/microsoft/WSL-DistroLauncher) that does exactly that as first command, but so far I haven't polished it for publishing yet. The diff is as follows:

```diff
diff --git DistroLauncher-Appx/DistroLauncher-Appx.vcxproj DistroLauncher-Appx/DistroLauncher-Appx.vcxproj
index 9b8a719..75bb013 100644
--- DistroLauncher-Appx/DistroLauncher-Appx.vcxproj
+++ DistroLauncher-Appx/DistroLauncher-Appx.vcxproj
@@ -10,7 +10,7 @@
     <WindowsTargetPlatformVersion>10.0</WindowsTargetPlatformVersion>
     <WindowsTargetPlatformMinVersion>10.0.18362.0</WindowsTargetPlatformMinVersion>
     <ApplicationTypeRevision>10.0</ApplicationTypeRevision>
-    <TargetName>mydistro</TargetName>
+    <TargetName>nixos</TargetName>
     <ProjectName>DistroLauncher-Appx</ProjectName>
     <AppxAutoIncrementPackageRevision>True</AppxAutoIncrementPackageRevision>
   </PropertyGroup>
@@ -78,11 +78,12 @@
     <PackageCertificateKeyFile>DistroLauncher-Appx_TemporaryKey.pfx</PackageCertificateKeyFile>
     <AppxAutoIncrementPackageRevision>True</AppxAutoIncrementPackageRevision>
     <AppxBundle>Always</AppxBundle>
     <GenerateAppInstallerFile>False</GenerateAppInstallerFile>
     <AppxBundlePlatforms>x64|arm64</AppxBundlePlatforms>
     <AppInstallerUpdateFrequency>1</AppInstallerUpdateFrequency>
     <AppInstallerCheckForUpdateFrequency>OnApplicationRun</AppInstallerCheckForUpdateFrequency>
+    <AppxPackageSigningEnabled>True</AppxPackageSigningEnabled>
   </PropertyGroup>
   <ItemDefinitionGroup>
     <Link>
diff --git DistroLauncher-Appx/MyDistro.appxmanifest DistroLauncher-Appx/MyDistro.appxmanifest
index f934c6c..9c5a740 100644
--- DistroLauncher-Appx/MyDistro.appxmanifest
+++ DistroLauncher-Appx/MyDistro.appxmanifest
@@ -1,9 +1,9 @@
 ∩╗┐<?xml version="1.0" encoding="utf-8"?>
 <Package xmlns="http://schemas.microsoft.com/appx/manifest/foundation/windows10" xmlns:mp="http://schemas.microsoft.com/appx/2014/phone/manifest" xmlns:uap="http://schemas.microsoft.com/appx/manifest/uap/windows10" xmlns:uap2="http://schemas.microsoft.com/appx/manifest/uap/windows10/2" xmlns:uap3="http://schemas.microsoft.com/appx/manifest/uap/windows10/3" xmlns:desktop="http://schemas.microsoft.com/appx/manifest/desktop/windows10" xmlns:rescap="http://schemas.microsoft.com/appx/manifest/foundation/windows10/restrictedcapabilities" IgnorableNamespaces="uap mp uap2 uap3 rescap desktop">
-  <Identity Name="WSL-DistroLauncher" Version="1.0.0.0" Publisher="CN=Windows Console Dev Team" ProcessorArchitecture="x64" />
+  <Identity Name="WSL-NixOS" Version="1.0.0.0" Publisher="CN=andy" ProcessorArchitecture="x64" />
   <mp:PhoneIdentity PhoneProductId="160867c6-4e75-4e36-85c6-1543de07d5f3" PhonePublisherId="00000000-0000-0000-0000-000000000000" />
   <Properties>
-    <DisplayName>WSL-DistroLauncher</DisplayName>
+    <DisplayName>WSL-NixOS</DisplayName>
     <PublisherDisplayName>Windows Console Dev Team</PublisherDisplayName>
     <Logo>Assets\StoreLogo.png</Logo>
   </Properties>
@@ -17,8 +17,8 @@
     <Resource Language="x-generate" />
   </Resources>
   <Applications>
-    <Application Id="mydistro" Executable="mydistro.exe" EntryPoint="Windows.FullTrustApplication">
-      <uap:VisualElements DisplayName="WSL-DistroLauncher" Description="My Own Distro Launcher" Square150x150Logo="Assets\Square150x150Logo.png" Square44x44Logo="Assets\Square44x44Logo.png" BackgroundColor="transparent">
+    <Application Id="mydistro" Executable="nixos.exe" EntryPoint="Windows.FullTrustApplication">
+      <uap:VisualElements DisplayName="WSL-NixOS" Description="NixOS on WSL" Square150x150Logo="Assets\Square150x150Logo.png" Square44x44Logo="Assets\Square44x44Logo.png" BackgroundColor="transparent">
         <uap:DefaultTile Wide310x150Logo="Assets\Wide310x150Logo.png" Square310x310Logo="Assets\LargeTile.png" Square71x71Logo="Assets\SmallTile.png">
         </uap:DefaultTile>
         <uap:SplashScreen Image="Assets\SplashScreen.png" />
@@ -26,10 +26,10 @@
       <Extensions>
         <uap3:Extension Category="windows.appExecutionAlias" Executable="mydistro.exe" EntryPoint="Windows.FullTrustApplication">
           <uap3:AppExecutionAlias>
-            <desktop:ExecutionAlias Alias="mydistro.exe" />
+            <desktop:ExecutionAlias Alias="nixos.exe" />
           </uap3:AppExecutionAlias>
         </uap3:Extension>
       </Extensions>
     </Application>
   </Applications>
-</Package>
\ No newline at end of file
+</Package>
diff --git DistroLauncher/DistributionInfo.cpp DistroLauncher/DistributionInfo.cpp
index 4742f59..a8c0098 100644
--- DistroLauncher/DistributionInfo.cpp
+++ DistroLauncher/DistributionInfo.cpp
@@ -7,33 +7,22 @@
 
 bool DistributionInfo::CreateUser(std::wstring_view userName)
 {
-    // Create the user account.
+    // Activate system profile
     DWORD exitCode;
-    std::wstring commandLine = L"/usr/sbin/adduser --quiet --gecos '' ";
-    commandLine += userName;
-    HRESULT hr = g_wslApi.WslLaunchInteractive(commandLine.c_str(), true, &exitCode);
+    std::wstring commandLine = L"/nix/var/nix/profiles/system/activate";
+    HRESULT hr = g_wslApi.WslLaunchInteractive(commandLine.c_str(), false, &exitCode);
     if ((FAILED(hr)) || (exitCode != 0)) {
         return false;
     }
 
-    // Add the user account to any relevant groups.
-    commandLine = L"/usr/sbin/usermod -aG adm,cdrom,sudo,dip,plugdev ";
-    commandLine += userName;
-    hr = g_wslApi.WslLaunchInteractive(commandLine.c_str(), true, &exitCode);
-    if ((FAILED(hr)) || (exitCode != 0)) {
-
-        // Delete the user if the group add command failed.
-        commandLine = L"/usr/sbin/deluser ";
-        commandLine += userName;
-        g_wslApi.WslLaunchInteractive(commandLine.c_str(), true, &exitCode);
-        return false;
-    }
-
     return true;
 }
 
 ULONG DistributionInfo::QueryUid(std::wstring_view userName)
 {
+	// XXX explain
+	return 0;
+
     // Create a pipe to read the output of the launched process.
     HANDLE readPipe;
     HANDLE writePipe;
@@ -41,7 +30,7 @@ ULONG DistributionInfo::QueryUid(std::wstring_view userName)
     ULONG uid = UID_INVALID;
     if (CreatePipe(&readPipe, &writePipe, &sa, 0)) {
         // Query the UID of the supplied username.
-        std::wstring command = L"/usr/bin/id -u ";
+        std::wstring command = L"/run/current-system/sw/bin/id -u ";
         command += userName;
         int returnValue = 0;
         HANDLE child;
diff --git DistroLauncher/DistributionInfo.h DistroLauncher/DistributionInfo.h
index cceb0e2..bdd02d8 100644
--- DistroLauncher/DistributionInfo.h
+++ DistroLauncher/DistributionInfo.h
@@ -13,10 +13,10 @@ namespace DistributionInfo
     //
     // WARNING: This value must not change between versions of your app,
     // otherwise users upgrading from older versions will see launch failures.
-    const std::wstring Name = L"MyDistribution";
+    const std::wstring Name = L"NixOS";
 
     // The title bar for the console window while the distribution is installing.
-    const std::wstring WindowTitle = L"My Distribution";
+    const std::wstring WindowTitle = L"NixOS on WSL";
 
     // Create and configure a user account.
     bool CreateUser(std::wstring_view userName);
```

**Comment by DanielFabian** [comment]
I've managed to it using your approach. It's true that it's a bit rough around the edges, though. Like it seems that it's created a user `andy` and it's got the password already set (?). At any rate, systemctl did work! So once you've got this in a shape that's a little bit less rough, I'm very happy to give it another go. For now, I think, I'll stay with home-manager on top of WSL Ubuntu. Unless you want some help?

**Comment by Trundle** [comment]
It creates a user that is defined in `configuration.nix`, which defaults to `andy`, without any password set. `configuration.nix` is what will end up under `/etc/nixos/configuration.nix`, i.e. NixOS's familiar configuration file. That means that a certain degree of customization is certainly expected, just like one would do when installing NixOS.

At this point, NixOS-WSL scratches my itch and I'm using it on a daily basis and it fulfills my needs. That means, without concrete improvement suggestions, I'm really sorry, but I don't see what I can do to improve NixOS-WLS's current shape. Note that my impression was that upstream's interest in WSL was rather non-existent, hence at this point I'm also not interested at putting more energy into this project. On the other hand, contributions welcome, so feel free to contribute.

**Comment by Trundle** [comment]
Also, please note that the concrete steps how to use this (`/nix/var/nix/profiles/system/activate`) are already mentioned in the README.

**Comment by izzues** [comment]
Sorry if this is not the right place to ask, but I have the same question as the issue title, just in a more basic level. The README is a little confusing, so I'm not sure where I should even begin. I'm trying to guess here and write a step-by-step list, so I would really appreciate if you could tell me if I'm getting all of it right:

* Have Nix already installed somewhere.
* Clone the repo.
* `cd` into the repo and run `nix-build -A system -A config.system.build.tarball ./nixos.nix` to obtain a tarball.
* Run `wsl --import <tarball>`? Is DistroLauncher optional?
* Run `/nix/var/nix/profiles/system/activate`
* Done? Any other steps needed (e.g. "go read Nix post-install docs")?

Thanks in advance!

**Comment by Kaeptenblaubaer** [comment]
@DanielFabian You can create /etc/nixos in Windows Explorer by opening 
`\\wsl$\NIXOS\etc\nixos`
and then insert your configuration.nix and hardware-configuration.nix
and execute  `nixos-rebuild switch`

**Comment by Kaeptenblaubaer** [comment]
I'm continuing to try to make this usable for NixOS/WSL-Newbies like me. I forked to

https://github.com/Kaeptenblaubaer/NixOS-WSL

**Comment by sonowz** [comment]
> I'm continuing to try to make this usable for NixOS/WSL-Newbies like me. I forked to
> 
> https://github.com/Kaeptenblaubaer/NixOS-WSL

I successfully ran NixOS 20.03 with your guide! 👍 
Except I had to add `export LC_ALL=$LANG` line in `activate.sh`, or I had a locale error..

Also, for whom wants minimal 'quirks' of WSL2, here is what I narrowed down:
```nix
let
  defaultUser = "%YOUR_USERNAME_HERE%";
  syschdemd = import ./%PATH_TO_THE_FILE%/syschdemd.nix { inherit lib pkgs config defaultUser; };
in
{
  boot.isContainer = true;
  users.users.root = {
    shell = "${syschdemd}/bin/syschdemd";
    # Otherwise WSL fails to login as root with "initgroups failed 5"
    extraGroups = [ "root" ];
  };

  # Described as "it should not be overwritten" in NixOS documentation,
  # but it's on /run per default and WSL mounts /run as a tmpfs, hence
  # hiding the wrappers.
  security.wrapperDir = "/wrappers";

  # Disable systemd units that don't make sense on WSL
  systemd.services."serial-getty@ttyS0".enable = false;
  systemd.services."serial-getty@hvc0".enable = false;
  systemd.services."getty@tty1".enable = false;
  systemd.services."autovt@".enable = false;
  systemd.services.firewall.enable = false;
  systemd.services.systemd-resolved.enable = false;
  systemd.services.systemd-udevd.enable = false;
  # Don't allow emergency mode, because we don't have a console.
  systemd.enableEmergencyMode = false;
}
```

I recommend running `sudo nixos-generate-config` after first boot and append these in `hardware-configuration.nix`. ([syschdemd.nix](https://github.com/Trundle/NixOS-WSL/blob/master/syschdemd.nix) and [syschdemd.sh](https://github.com/Trundle/NixOS-WSL/blob/master/syschdemd.sh) from this repo is also required)
Then you can write down your `configuration.nix` whatever you want!

**Comment by alexvorobiev** [comment]
I have followed the guide but got Unspecified error trying to import. I've opened https://github.com/Trundle/NixOS-WSL/issues/4.

**Comment by Trundle** [comment]
I just merged #7 which adds a quick start guide to the README. Please feel free to reopen (or come up with a merge request) if things are still unclear.
